PictureApp/Controller/PicturesViewController.swift

```diff
@@ -0,0 +1,306 @@
+        filterVc.delegate = self
+        filterVc.title = "Select a Category"
+        
+        navigationController?.pushViewController(filterVc, animated: true)
+        
+        
+    }
+    
+    
+    func setupCollectionView(){
+        
+        collectionView.register(GalleryCollectionViewCell.self, forCellWithReuseIdentifier: GalleryCollectionViewCell.identifier)
+        collectionView.dataSource = self
+        collectionView.delegate = self
+        setupCollectionViewLayout()
+        view.addSubview(collectionView)
+        let margins = view.layoutMarginsGuide
+        collectionView.translatesAutoresizingMaskIntoConstraints = false
+        collectionView.leadingAnchor.constraint(equalTo: margins.leadingAnchor).isActive = true
+        collectionView.bottomAnchor.constraint(equalTo: margins.bottomAnchor).isActive = true
+        collectionView.trailingAnchor.constraint(equalTo: margins.trailingAnchor).isActive = true
+        collectionView.topAnchor.constraint(equalTo: margins.topAnchor).isActive = true
+        
+        
+    }
+    
+     
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        let buttonOptions = UIButton(type: .contactAdd)
+        buttonOptions.addTarget(self, action: #selector(showOptions(_:)), for: .touchUpInside)
+        navigationItem.rightBarButtonItem = .init(customView: buttonOptions)
+        setupCollectionView()
+
+        fetchHits(order: orderForCategories,selectedCategory: ["music"],currentPage: currentPage)
+        
+    }
+
+}
+    
+extension PicturesViewController : UICollectionViewDelegate,UICollectionViewDataSource {
+    
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return hits.count
+    }
+  
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: GalleryCollectionViewCell.identifier, for: indexPath) as! GalleryCollectionViewCell
+        
+        let tagText = hits[indexPath.item].tags
+      
+        let url = URL(string: hits[indexPath.row].image)!
+        ImageLoader(with: URLHTTPClient(session: URLSession.shared)).load(from: url) { (image) in
+            DispatchQueue.main.async {
+                guard collectionView.indexPath(for: cell) == indexPath else {
+
+                    return }
+
+                cell.imageView.image = image
+                cell.tagsLabel.text = tagText
+        
+            }
+        }
+        
+        return cell
+    }
+    
+    
+    //MARK:Pagination
+    func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
+        if hits.count - 1 < totalHits && indexPath.row == hits.count - 1 {
+            currentPage = currentPage + 1
```

Comment:
In case of error occurs in the request you have already incremented the page.

So for example:

You fetched first page, but got error on the second you will try to fetch third page instead of second.

Reply:
Fixed